### PR TITLE
Add build script invalidation and snapshotting logic

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Invalidate the build script whenever the set of `builders` is changed. Use a snapshotted build script instead of running from source. Reduces daemon start up time to 0.9 seconds from about 10 seconds.

## Related Issues

https://github.com/flutter/flutter/issues/27141

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_runner/build_runner.dart [hunk 1]
```diff
@@ -127,57 +126,74 @@ class BuildRunner extends CodeGenerator {
...
+    final Directory generatedDirectory = fs.directory(fs.path.join(flutterProject.dartTool.path, 'flutter_tool'));
+    final File buildScript = entrypointDirectory.childFile('build.dart');
+    final File buildSnapshot = entrypointDirectory.childFile('build.dart.snapshot');
+    final File scriptIdFile = entrypointDirectory.childFile('id');
+    final File syntheticPubspec = generatedDirectory.childFile('pubspec.yaml');
+
+    // Check if contents of builders changed. If so, invalidate build script
+    // and regnerate.
+    final YamlMap builders = await flutterProject.builders;
+    final String newContents = _produceScriptId(builders);
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/build_runner.dart]: This code deals with a large number of files. I think a more specific name for this variable could improve readability. "contents" sounds too generic. It's unclear which file it comes from.

File: packages/flutter_tools/lib/src/build_runner/build_runner.dart [hunk 2]
```diff
@@ -279,3 +293,14 @@ class _BuildRunnerCodegenDaemon implements CodegenDaemon {
...
+
+// Sorts the builders by name and produces a hashcode of the resulting iterable.
+String _produceScriptId(YamlMap builders) {
+  if (builders == null || builders.isEmpty) {
+    return '0';
+  }
+  final List<String> orderedBuilders = builders.keys
+    .cast<String>()
+    .toList()..sort();
+  return hashObjects(orderedBuilders).toString();
```
Comment (yjbanov) [inline packages/flutter_tools/lib/src/build_runner/build_runner.dart]: Is it possible to reuse https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/fingerprint.dart for this? Hash codes are not very reliable w.r.t. collisions.

**Comment by jonahwilliams** [comment]
Replaced hashcodes with md5 and renamed locals

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**